src/main/scala/beam/analysis/RealizedModeChoiceWriter.scala

```diff
@@ -0,0 +1,86 @@
+package beam.analysis
+
+import java.util
+
+import beam.agentsim.events.handling.{BeamEventsLoggingSettings, BeamEventsWriterCSV}
+import beam.agentsim.events.{ModeChoiceEvent, ReplanningEvent}
+import beam.sim.BeamServices
+import org.matsim.api.core.v01.Id
+import org.matsim.api.core.v01.events.Event
+import org.matsim.api.core.v01.population.Person
+import org.matsim.core.controler.events.{IterationEndsEvent, IterationStartsEvent}
+import org.matsim.core.controler.listener.{IterationEndsListener, IterationStartsListener}
+import org.matsim.core.events.handler.BasicEventHandler
+
+import scala.collection.mutable
+
+/**
+  * Listens for [[ModeChoiceEvent]] and [[ReplanningEvent]] to calculate
+  * realized mode choices.
+  * Writes realizedModeChoice.csv.gz to iteration directory.
+  */
+class RealizedModeChoiceWriter(beamServices: BeamServices)
+    extends BasicEventHandler
+    with IterationEndsListener
+    with IterationStartsListener {
+
+  private var csvWriter: BeamEventsWriterCSV = _
+  private var csvFilePath: String = _
+
+  /** Collects a previous ModeChoiceEvent for each person */
+  private val personIdPrevMCE = mutable.Map.empty[Id[Person], Option[ModeChoiceEvent]]
+
+  override def handleEvent(event: Event): Unit = {
+    event match {
+      case modeChoiceEvent: ModeChoiceEvent =>
+        val personId = modeChoiceEvent.personId
+        val prevMCE = personIdPrevMCE.get(personId).flatten
+
+        prevMCE match {
+          case Some(realizedMCE @ _) =>
+            // second (current) ModeChoiceEvent makes previous one "realized"
+            csvWriter.writeEvent(realizedMCE)
+            personIdPrevMCE(personId) = None
+
+          case _ =>
+            personIdPrevMCE(personId) = Some(modeChoiceEvent)
+        }
+
+      case replanningEvent: ReplanningEvent =>
+        // ReplanningEvent discards previous history
+        personIdPrevMCE(replanningEvent.getPersonId) = None
+
+      case _ =>
+    }
+  }
+
+  override def notifyIterationStarts(event: IterationStartsEvent): Unit = {
+    csvFilePath = beamServices.matsimServices.getControlerIO.getIterationFilename(
+      event.getIteration,
+      "realizedModeChoice.csv.gz"
+    )
+
+    // writeHeader is invoked in constructor
+    csvWriter = new BeamEventsWriterCSV(
```

Comment:
possibility  for resource leak , close csv writer for any exceptions

Reply:
This is the last thing done in this method - and closing it on exceptions outside of this won't make a difference. So, the only place needed would be `notifyIterationEnds` as far as I can tell